cli/src/commands/simplify_parents.rs

```diff
@@ -0,0 +1,119 @@
+/// ancestor of B, A will be rewritten to have only B as a parent instead of
+/// B+C.
+#[derive(clap::Args, Clone, Debug)]
+#[command(group = clap::ArgGroup::new("revision-args").multiple(true).required(true))]
+pub(crate) struct SimplifyParentsArgs {
+    /// Simplify specified revision(s) together with their trees of descendants
+    /// (can be repeated)
+    #[arg(long, short, group = "revision-args")]
+    sources: Vec<RevisionArg>,
+    /// Simplify specified revision(s) (can be repeated)
+    #[arg(long, short, group = "revision-args")]
+    revisions: Vec<RevisionArg>,
+}
+
+pub(crate) fn cmd_simplify_parents(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &SimplifyParentsArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let revs = RevsetExpression::descendants(
+        workspace_command
+            .parse_union_revsets(&args.sources)?
+            .expression(),
+    )
+    .union(
+        workspace_command
+            .parse_union_revsets(&args.revisions)?
+            .expression(),
+    );
+    let commit_ids = workspace_command
+        .attach_revset_evaluator(revs)
+        .evaluate_to_commit_ids()?
+        .collect_vec();
+    workspace_command.check_rewritable(&commit_ids)?;
+    let num_orig_commits = commit_ids.len();
+
+    let mut tx = workspace_command.start_transaction();
+    let mut stats = SimplifyStats::default();
+    tx.repo_mut()
+        .transform_descendants(command.settings(), commit_ids, |rewriter| {
+            simplify_commit_parents(command.settings(), rewriter, &mut stats)
+        })?;
+
+    if let Some(mut formatter) = ui.status_formatter() {
+        if !stats.is_empty() {
+            writeln!(
+                formatter,
+                "Removed {} edges from {} out of {} commits.",
+                stats.edges, stats.commits, num_orig_commits
+            )?;
+        }
+    }
+    tx.finish(ui, format!("simplify {} commits", num_orig_commits))?;
+
+    Ok(())
+}
+
+#[derive(Default)]
+struct SimplifyStats {
+    commits: usize,
+    edges: usize,
+}
+
+impl SimplifyStats {
+    fn for_commit(edges: usize) -> Self {
+        Self { commits: 1, edges }
+    }
+
+    fn add(&mut self, other: SimplifyStats) {
+        self.commits += other.commits;
+        self.edges += other.edges;
+    }
+
+    fn is_empty(&self) -> bool {
+        self.commits == 0 && self.edges == 0
+    }
+}
+
+fn simplify_commit_parents(
+    settings: &UserSettings,
+    mut rewriter: CommitRewriter,
+    stats: &mut SimplifyStats,
+) -> BackendResult<()> {
+    if rewriter.old_commit().parent_ids().len() <= 1 {
+        return Ok(());
+    }
+
+    let num_old_heads = rewriter.new_parents().len();
+    rewriter.simplify_ancestor_merge();
+    let num_new_heads = rewriter.new_parents().len();
+
+    if num_new_heads < num_old_heads {
+        rewriter.reparent(settings)?.write()?;
```

Comment:
probably better to do `.reparent()` if `.parents_changed()` so the descendants won't be rebased later by `tx.finish()`.